Implement Strings, newlines and positional edits using StringFormat

gptsarthak opened the pull request:
This PR builds up on #1829 and implement Strings, newlines and positional edits using ```StringFormat```.
@certik Please review my latest commit. If its alright then I'll continue adding other formatting types.

The following produces the expected output:
```fortran
program format_04
print *, "ok", "b"
print '(a,a)', "ok", "b"
print '("Success!",/,10X,A6,"World!")',"Hello 123"
end program
```
Output:
```console
ok b
okb
Success!
          Hello World!
```

certik commented:
@gptsarthak yes, that's very good. Keep working in this PR and add more types in the formatter. I am working on getting it to pass all the CI tests.

certik commented:
@gptsarthak I finally got the initial implementation fully working in https://github.com/lfortran/lfortran/pull/1828, all tests pass and I merged it. I then rebased this PR on top of the latest main, resolved conflicts. All tests should pass now. Make sure you pull this PR to your local branch, like this:
```
git checkout main
git fetch sarthak +format2:format2
git checkout sarthak
```
Where `sarthak` is your remote for your fork. Then you can add more commits with more features.

gptsarthak commented:
Implemented Repeat Count```(5A4)```, Integers ```(Iw.d)``` and Decimals ```(Dw.d)```
Implemented Exponent partially ```(Ew.d)```

TODO:
- Implement nested formats ```(6(2X,E11.5))```
- Fix bug: Only ```Real*8``` values are recognized by variadic arguments, ```double val  = va_arg(args,double)``` does not work for simply ```real``` types. 
- Implement Scaling Factor ```(1P)```
- Implement ```(Ew.dEe)``` but might be low priority since Scipy does not use it
- Thoroughly test everything

After these we will implement the fixed-form ```FORMAT``` statement to use ```StringFormat```.

certik commented:
Excellent progress! At the CI there is a failure on macOS, it seems like some kind of a segfault or a dangling pointer, so we'll have to fix that.

gptsarthak commented:
@certik how should I try to fix it? Is there a way I can test the macOS workflow on my pc? 
Will dynamically allocating the strings which are currently defined as 
```char val_str[64]```  and  ```char formatted_value[64] = ""```  fix the issue?

Shaikh-Ubaid commented:
>how should I try to fix it? Is there a way I can test the macOS workflow on my pc?

I think macos builds in release mode. I would suggest to try building in release mode and see if you are able to reproduce the error. (Maybe the bug is present in the release build?)

gptsarthak commented:
Thank you so much @Shaikh-Ubaid , I have found the error in release build and will push the fix in the next commit.

certik commented on src/libasr/runtime/lfortran_intrinsics.c:
```diff
@@ -113,17 +113,297 @@ LFORTRAN_API void _lfortran_printf(const char* format, ...)
+    return substr;
+}
+
+char* appendToString(char* str, const char* append) {
```
Let's use this convention, consistent with the rest of the file:
```suggestion
char* append_to_string(char* str, const char* append) {
```

certik commented:
Looks like you figured it out. Perfect, it looks very good in terms of features. 

Once this is in, we can come back and clean it up. We should introduce a custom Str struct and associated functions, similar to the [Vec](https://github.com/lfortran/lfortran/blob/a7cd2e6bb98de767f8f787d4f068d058c32db083/src/libasr/containers.h#L42) class, but acting like `std::string`, it will have a length and capacity. So appending at the end will become cheap, no reallocation necessary. We'll do it later.

certik commented:
Very good. Let me know when you want this to be reviewed & merged.
